devito/finite_difference.py

```diff
@@ -221,3 +224,196 @@ def second_cross_derivative(function, dims, order):
+            default_offset = dict((k, 0) for k in self[0][1].keys())
+            if isinstance(other, Function):
+                new = sparse_fd_list([(other, default_offset)])
+            else:
+                new = sparse_fd_list([(o, default_offset) for o in other.args])
+            return sparse_fd_list(super(sparse_fd_list, self).__add__(new)).simplify()
+
+    def __iadd__(self, other):
+        """
+        Overload addition of two fd list or addition with a constant
+        The constant is either a symbolic expression or function
+        """
+        if isinstance(other, sparse_fd_list):
+            return sparse_fd_list(super(sparse_fd_list, self).__iadd__(other)).simplify()
+        else:
+            default_offset = dict((k, 0) for k in self[0][1].keys())
+            if isinstance(other, Function):
+                new = sparse_fd_list([(other, default_offset)])
+            else:
+                new = sparse_fd_list([(o, default_offset) for o in other.args])
+            return sparse_fd_list(super(sparse_fd_list, self).__iadd__(new)).simplify()
+
+    def __mul__(self, other_list):
+        """
+        Overload of the multiplication by a constant. FD offsets stay unchanged and
+        weights are multiplied by the constant (symbolic or value)
+        """
+        return sparse_fd_list([(i[0] * other_list, i[1]) for i in self]).simplify()
+
+    def __rmul__(self, constant):
+        """
+        Overload of the multiplication by a constant. FD offsets stay unchanged and
+        weights are multiplied by the constant (symbolic or value)
+        """
+        return sparse_fd_list([(i[0] * constant, i[1]) for i in self]).simplify()
+
+    def __floordiv__(self, constant):
+        """
+        Overload of the division by a constant. FD offsets stay unchanged and
+        weights are divided by the constant (symbolic or value)
+        """
+        return sparse_fd_list([(i[0] / constant, i[1]) for i in self]).simplify()
+
+    def __truediv__(self, constant):
+        """
+        Overload of the division by a constant. FD offsets stay unchanged and
+        weights are divided by the constant (symbolic or value)
+        """
+        return sparse_fd_list([(i[0] / constant, i[1]) for i in self]).simplify()
+
+    def simplify(self):
+        """
+        Remove 0 coefficient terms and sums common fd offset terms
+        """
+        offsets = self.unique_offset()
+        out = []
+        for off in offsets:
+            w = 0
+            for v, offs in self:
+                w += v if offs == off else 0
+            out += [(w, off)]
+        return sparse_fd_list([f for f in out if f[0] != 0])
+
+    def unique_offset(self):
+        """
+        Utility funstion that returns the list of unique FD offset for factorization
+        """
+        unique = []
+        for _, off in self:
+            if off not in unique:
+                unique.append(off)
+        return unique
+
+
+def sum_dict(d1, d2):
```

Comment:
better, one-line alternatives here: https://stackoverflow.com/questions/10461531/merge-and-sum-of-two-dictionaries